### PR TITLE
Support windows where mkstemp and ftruncate doesn't exist

### PR DESCRIPTION
**Pull request by isuruf**
Fixes https://github.com/flang-compiler/f18/issues/1040

### CODE REVIEW

File: runtime/file.cpp [hunk 1]
```diff
@@ -22,6 +28,35 @@ void OpenFile::set_path(OwningPtr<char> &&path, std::size_t bytes) {
...
+  int nBufferLength{MAX_PATH-14};
+  nBufferLength = ::GetTempPathA(nBufferLength, tempDirName);
+  if (nBufferLength > MAX_PATH-14) {
+    return -1;
+  }
+  if (::GetTempFileNameA(tempDirName, "f18", uUnique, tempFileName) == 0) {
+    return -1;
+  }
+  int fd{::_open(tempFileName, _O_CREAT | _O_TEMPORARY, _S_IREAD | _S_IWRITE)};
+  if (fd < 0) {
```
Comment (klausler) [inline runtime/file.cpp]: This part of the code is identical to the code in the other branch and can be moved after the `endif`.

File: runtime/file.cpp [hunk 2]
```diff
@@ -22,6 +28,35 @@ void OpenFile::set_path(OwningPtr<char> &&path, std::size_t bytes) {
...
+#ifdef _WIN32
+  const unsigned int uUnique{0};
+  char tempDirName[MAX_PATH-14];
+  char tempFileName[MAX_PATH];
+  int nBufferLength{MAX_PATH-14};
+  nBufferLength = ::GetTempPathA(nBufferLength, tempDirName);
+  if (nBufferLength > MAX_PATH-14) {
+    return -1;
+  }
+  if (::GetTempFileNameA(tempDirName, "f18", uUnique, tempFileName) == 0) {
```
Comment (klausler) [inline runtime/file.cpp]: Somebody might complain about the use of the name "f18" here.

File: runtime/file.cpp [hunk 3]
```diff
@@ -22,6 +28,35 @@ void OpenFile::set_path(OwningPtr<char> &&path, std::size_t bytes) {
   pathLength_ = bytes;
 }
 
+static int openfile_mkstemp(IoErrorHandler &handler) {
+#ifdef _WIN32
+  const unsigned int uUnique{0};
+  char tempDirName[MAX_PATH-14];
```
Comment (klausler) [inline runtime/file.cpp]: Why the adjustment by 14 here and below?  Is this a magic constant that should be named?

File: runtime/file.cpp [hunk 4]
```diff
@@ -22,6 +28,35 @@ void OpenFile::set_path(OwningPtr<char> &&path, std::size_t bytes) {
...
+  char tempFileName[MAX_PATH];
+  int nBufferLength{MAX_PATH-14};
```
Comment (klausler) [inline runtime/file.cpp]: `sizeof tempDirName` might be more robust.

File: runtime/file.cpp [hunk 5]
```diff
@@ -13,7 +13,13 @@
 #include <cstring>
 #include <fcntl.h>
 #include <stdlib.h>
+#ifndef _WIN32
```
Comment (klausler) [inline runtime/file.cpp]: Elsewhere in this code, you've organized the dependent variants with `ifdef _WIN32` and put the Windows code first.

File: runtime/file.cpp (hunk 3 above)
Comment (isuruf) [reply runtime/file.cpp]: The input to GetTempFileNameA should be less than or equal to MAX_PATH-14. Docs at https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettempfilenamea

Comment (klausler) [reply runtime/file.cpp]: A comment to that effect would reduce the mystery.

File: runtime/file.cpp (hunk 1 above)
Comment (isuruf) [reply runtime/file.cpp]: There's a unlink call in the non-windows code path which make it ugly. I pushed the change anyways.

**Comment by isuruf** [comment]
Should I squash the commits now or after a review?

**Comment by klausler** [comment]
> Should I squash the commits now or after a review?

Generally, it's best to squash after review, right before merging.

**Comment by isuruf** [comment]
Squashed and tests pass now.

**Pull request merged**